fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/cursors/SizeStatisticsCollectorCursor.java

```diff
@@ -0,0 +1,527 @@
+        }
+    }
+
+    @Nonnull
+    @Override
+    public Executor getExecutor() {
+        return context.getExecutor();
+    }
+
+    @Nonnull
+    @Override
+    public boolean accept(@Nonnull RecordCursorVisitor visitor) {
+        visitor.visitEnter(this);
+        return visitor.visitLeave(this);
+    }
+
+    @Nonnull
+    private SizeStatisticsResults emitFinalStats() {
+        return sizeStatisticsResults;
+    }
+
+    @Nonnull
+    @Override
+    @Deprecated
+    public CompletableFuture<Boolean> onHasNext() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Nullable
+    @Override
+    @Deprecated
+    public SizeStatisticsResults next() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Nullable
+    @Override
+    @Deprecated
+    public byte[] getContinuation() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Nonnull
+    @Override
+    @Deprecated
+    public RecordCursor.NoNextReason getNoNextReason() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void close() {
+        if (hasNextFuture != null) {
+            hasNextFuture.cancel(false);
+        }
+    }
+
+    //form a continuation that allows us to restart statistics aggregation from where we left off
+    private class SizeStatisticsCollectorCursorContinuation implements RecordCursorContinuation {
+        @Nonnull
+        private final RecordCursorResult<KeyValue> currentKvResult;
+        @Nonnull
+        private final Function<byte[], RecordCursorProto.SizeStatisticsContinuation> continuationFunction;
+        @Nonnull
+        private final boolean isEnd = false;
+        @Nullable
+        private byte[] cachedBytes;
+
+        private SizeStatisticsCollectorCursorContinuation(RecordCursorResult<KeyValue> currentKvResult) {
+            this.cachedBytes = null;
+            this.currentKvResult = currentKvResult;
+
+            //defer forming bytes until requested
+            continuationFunction = b -> {
+                RecordCursorProto.SizeStatisticsPartialResults.Builder pr;
+                if (finalResultsEmitted == false) {
+                    //more to aggregate put partial results in continuation
+                    pr = RecordCursorProto.SizeStatisticsPartialResults.newBuilder().setKeyCount(sizeStatisticsResults.keyCount).setKeySize(sizeStatisticsResults.keySize).setMaxKeySize(sizeStatisticsResults.maxKeySize).setValueSize(sizeStatisticsResults.valueSize).setMaxValueSize(sizeStatisticsResults.maxValueSize);
+                    List<Long> bucketList = Arrays.stream(sizeStatisticsResults.getSizeBuckets()).boxed().collect(Collectors.toList());
+                    pr.addAllSizeBuckets(bucketList);
+                    return RecordCursorProto.SizeStatisticsContinuation.newBuilder().setPartialResults(pr.build()).setContinuation(ByteString.copyFrom(b)).build();
```

Comment:
I believe the contract of `RecordCursorContinuation`s is that they should include only immutable state. As a result, we might need to eagerly create at least the `SizeStatisticsPartialResults` message.

Reply:
Not following the concern here. The continuation is still an opaque byte array. Please advise.

Reply:
Right, once the byte array has been materialized, this won't change, which is fine, but the contract is that this should never change its contents even as other things change (or, in other words, the continuation should only have immutable members). The continuation (before it's been byte-ified) relies on the mutable `SizeStatisticsResults` object, so it could change if someone, like, mutated that object.

Reply:
Makes sense. Will cache a copy.